vermouth/processors/locate_charge_dummies.py

```diff
@@ -0,0 +1,121 @@
+
+
+def colinear_pair():
+    # We want to create a randomly oriented vector of norm 2, and center it on
+    # the origin.
+    vector = np.random.rand(3)
+    vector /= np.linalg.norm(vector)
+    vector *= 2
+    points = np.stack([np.zeros((3, )), vector])
+    points -= points / 2
+    return points
+
+
+def find_anchor(molecule, node_key, attribute_tag=DEFAULT_DUMMY_ATTRIBUTE):
+    dummy = molecule.nodes[node_key]
+    if dummy.get(attribute_tag, None) is None:
+        msg = 'Node "{}" is not a charge dummy. Check the "{}" node attribute.'
+        raise ValueError(msg.format(node_key, attribute_tag))
+
+    # There should be only one anchor, and that anchor is the (hopefully) only
+    # neighbor that is not a dummy.
+    potential_anchors = [
+        neighbor
+        for neighbor in molecule.neighbors(node_key)
+        if molecule.nodes[neighbor].get(attribute_tag, None) is None
+    ]
+    if not potential_anchors:
+        raise ValueError('No anchor found for dummy bead "{}": {}.'
+                         .format(node_key, molecule.nodes[node_key]))
+    elif len(potential_anchors) > 1:
+        raise ValueError('Too many potential anchors found for dummy "{}" ({} found).'
+                         .format(node_key, len(potential_anchors)))
+
+    return potential_anchors[0]
+
+
+def locate_dummy(molecule, anchor_key, dummy_keys, attribute_tag=DEFAULT_DUMMY_ATTRIBUTE):
+    
+    anchor_position = molecule.nodes[anchor_key].get('position')
+    if anchor_position is None:
+        msg = 'The anchor of the "{}" dummy ("{}") does not have a position.'
+        raise ValueError(msg.format(node_key, anchor_position[0]))
+
+    distances = []
+    distance_error_keys = []
+    for dummy_key in dummy_keys:
+        try:
+            distances.append(float(molecule.nodes[dummy_key].get(attribute_tag)))
+        except ValueError:
+            distance_error_keys.append(dummy_key)
+    if distance_error_keys:
+        msg = ('The following charge dummies have an invalid for their {} ',
+               'attribute: {}. The values have to be numbers.'
+                .format(attribute_tag, ', '.join(distance_error_keys)))
+        raise ValueError(msg)
+    distances = np.array(distances)
+
+    if len(dummy_keys) == 2:
+        points = colinear_pair()
+    else:
+        points = fibonacci_sphere(len(dummy_keys))
+    points *= distances[:, None]
+
+    for dummy_key, position in zip(dummy_keys, points):
+        molecule.nodes[dummy_key]['position'] = position + anchor_position
+
+
+def locate_all_dummies(molecule, attribute_tag=DEFAULT_DUMMY_ATTRIBUTE):
+    dummies = [
+        (find_anchor(molecule, dummy_key, attribute_tag), dummy_key)
+        for dummy_key in molecule.nodes
+        if molecule.nodes[dummy_key].get(attribute_tag, None) is not None
+    ]
+    dummies.sort()  # Sort primarily on the anchor key as it appears first
+
+    grouped_by_anchor = itertools.groupby(dummies, key=operator.itemgetter(0))
```

Comment:
Are you sure you need the `itemgetter`?

Reply:
I could use lambda. Without a key, `groupby` use the full object while I want to only use the first element of a tuple of size 2.

Reply:
That I see. Does using the full tuple produce a qualitatively different result though? I do prefer the itemgetter over a lambda.

Reply:
It should. The list `dummies` has the form` [(anchor0, dummy0), (anchor0, dummy1), (anchor1, dummy2), ...]` and I want to group the dummies by their anchors. Since ll the dummies are different, grouping on the full tuple will keep all the element separated.